incubator/hnc/pkg/reconcilers/hnc_config.go

```diff
@@ -136,28 +138,52 @@ func (r *ConfigReconciler) createObjectReconcilers(inst *api.HNCConfiguration) {
 	// from the forest for the object reconciliation, after we create ObjectReconcilers but before we write the
 	// ObjectReconcilers to the types list. As a result, objects of the newly added types might not be propagated correctly
 	// using the latest forest structure.
+	//
+	// For syncing an object reconciler, the mutex is guarding the read access to the `namespaces` field in the forest. The
+	// `namespaces` field is a shared resource between various reconcilers.
 	r.Forest.Lock()
 	defer r.Forest.Unlock()
 
 	for _, t := range inst.Spec.Types {
 		gvk := schema.FromAPIVersionAndKind(t.APIVersion, t.Kind)
-		if r.Forest.HasTypeSyncer(gvk) {
-			continue
+		mode := r.getMode(t)
+		if ts := r.Forest.GetTypeSyncer(gvk); ts != nil {
+			r.syncObjectReconciler(ctx, mode, ts)
+		} else {
+			r.createObjectReconciler(gvk, mode, inst)
+		}
+	}
+}
+
+// syncObjectReconciler syncs the mode of an ObjectReconciler. It also updates corresponding objects, if needed.
+func (r *ConfigReconciler) syncObjectReconciler(ctx context.Context, mode api.SynchronizationMode, ts forest.TypeSyncer) {
+	r.Log.Info("Syncing an object reconciler.", "gvk", ts.GetGVK())
+
+	oldMode := ts.GetMode()
+	if mode != oldMode {
+		r.Log.Info("Changing Mode of the object reconciler", "old Mode", oldMode, "new Mode", mode)
+		ts.SetMode(mode)
+		// If we change from "ignore" mode to other modes (e.g., "propagate"), we need to update objects in the cluster
+		// (e.g., propagate existing objects).
+		if oldMode == api.Ignore {
+			err := ts.SyncCluster(ctx, r.Log)
+			if err != nil {
+				r.Log.Error(err, "Error while trying to sync ObjectReconciler", "gvk", ts.GetGVK())
+			}
 		}
-		r.createObjectReconciler(gvk, inst)
 	}
 }
 
 // createObjectReconciler creates an ObjectReconciler for the given GVK and informs forest about the reconciler.
-// TODO: May need to pass in spec instead to provide information about Mode to the ObjectReconciler in future.
-func (r *ConfigReconciler) createObjectReconciler(gvk schema.GroupVersionKind, inst *api.HNCConfiguration) {
-	r.Log.Info("Creating an object reconciler.", "GVK", gvk)
+func (r *ConfigReconciler) createObjectReconciler(gvk schema.GroupVersionKind, mode api.SynchronizationMode, inst *api.HNCConfiguration) {
+	r.Log.Info("Creating an object reconciler.", "gvk", gvk, "Mode", mode)
```

Comment:
Nit: no period, also lowercase `"Mode"`

Reply:
Removed period and changed to lowercase.